x/oracle/keeper/keeper.go

```diff
@@ -110,16 +107,16 @@ func (k Keeper) GetExchangeRateBase(ctx sdk.Context, denom string) (sdk.Dec, err
 	}
 
 	powerReduction := ten.Power(exponent)
-	return exchangeRate.Quo(powerReduction), nil
+	return exchangeRate.Rate.Quo(powerReduction), nil
 }
 
 // SetExchangeRate sets the consensus exchange rate of USD denominated in the
 // denom asset to the store.
-func (k Keeper) SetExchangeRate(ctx sdk.Context, denom string, exchangeRate sdk.Dec) {
-	store := ctx.KVStore(k.storeKey)
-	bz := k.cdc.MustMarshal(&sdk.DecProto{Dec: exchangeRate})
-	denom = strings.ToUpper(denom)
-	store.Set(types.KeyExchangeRate(denom), bz)
+func (k Keeper) SetExchangeRate(ctx sdk.Context, denom string, rate sdk.Dec) {
+	key := types.KeyExchangeRate(denom)
+	val := types.ExchangeRate{Rate: rate, Timestamp: ctx.BlockTime()}
```

Comment:
ctx.BlockTime() is still vulnerable to odd bugs, such as having zero value at times (which is why I store `LastInterestTime` and `LastRewardTime` with a safety layer in `leverage` and `incentive` modules)

Reply:
we have run into bugs on mainnet and testnets where ctx.BlockTime() has been wildly inaccurate - either zero or years in the past - for a single block. That's why I have "ignore zero / negative / decrease block time" when doing `LastInterestTime` for example.

Seems to come from tebndermint layer so we can't fix it